[TechRaj] iP

TechRaj opened the pull request:
## ProYapper

> “Are you a Yapper? And is your time management really bad? Don't worry! We have something that is better than you in both aspects!” – [TechRaj](https://youtu.be/dQw4w9WgXcQ)

**Introducing, ProYapper!** 🎉

ProYapper helps you:

- Organise your tasks
- Is easy to use
- Can Yap like a pro!

### Getting Started

All you need to do is:

1. [Download it from here](#). 
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉  
   
And it is **FREE!**

### Features

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

---

**ProYapper**: *Because even a Yapper needs some organization.* 😄

TechRaj committed:
Rename, Greet, Exit

TechRaj committed:
Echo

TechRaj committed:
Add, List

TechRaj committed:
Mark as Done

TechRaj committed:
ToDos, Events, Deadlines

TechRaj committed:
Automated Text UI Testing

TechRaj committed:
Handle Errors

TechRaj committed:
delete

TechRaj committed:
Save

TechRaj committed:
Tidy up code

TechRaj committed:
Dates and Times

TechRaj committed:
Make the code more OOP

TechRaj committed:
Divide classes into packages

TechRaj committed:
Add JavaDoc comments

TechRaj committed:
Tweak the code to comply with a coding standard

TechRaj committed:
Give users a way to find a task by searching for a keyword.

TechRaj committed:
add gradle support

TechRaj committed:
Clean up code

TechRaj committed:
Add GUI to chatbot using JavaFX

mattisongjj commented on src/main/java/commands/Command.java:
```diff
@@ -0,0 +1,21 @@
+     * @param ui the user interface to interact with the user
+     * @param storage the storage handler for loading and saving tasks
+     */
+    public abstract void execute(TaskList taskList, Ui ui, Storage storage);
```
I like the use of an abstract Command class that encapsulates command logic

mattisongjj replied:
Perhaps the commands can be further extracted out to more generic commands. For example a CreateCommand, DeleteCommand, EditCommand etc.

mattisongjj commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,123 @@
+ * The {@code Storage} class handles reading tasks from a file and saving tasks back to a file.
+ * This class ensures that task data is persistent between application runs.
+ */
+public class Storage {
```
Storage class perfectly encapsulates the fetching and storing of task data in a given file path

mattisongjj commented on src/main/java/commands/InvalidCommand.java:
```diff
@@ -0,0 +1,35 @@
+ * Represents a command that is invalid or unrecognized by the application.
+ * This command displays an error message to the user, indicating that the input was not valid.
+ */
+public class InvalidCommand extends Command {
```
Any reason for handling invalid commands this way, rather than creating an exception class?

mattisongjj commented on src/main/java/main/Parser.java:
```diff
@@ -0,0 +1,187 @@
+        String[] parts = userInput.split(" ", 2);
+        String commandWord = parts[0];
+
+        switch (commandWord) {
```
I like how clean this part of the code is :)

mattisongjj reviewed (approved):
LGTM

sannie-beep commented on src/main/java/commands/FindCommand.java:
```diff
@@ -0,0 +1,26 @@
+import ui.Ui;
+import tasks.Task;
+
+public class FindCommand extends Command {
```
Perhaps you forgot to include javadocs for this? But overall, nice job abstracting into different command classes! :)

sannie-beep commented on src/main/java/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,54 @@
+            Task newTask = new Deadline(this.taskDescription, this.dueWhen);
+            taskList.addTask(newTask);
+            int numTasks = taskList.size();
+            ui.showMessage("Got it. I've added this task:");
```
I noticed that many of the execute methods in various command classes for adding and removing tasks have the same pattern for the UI message displayed. Perhaps you could consider adding a method in the UI class like showTaskAddedMessage() or showTaskRemovedMessage() to abstract out this process?

sannie-beep commented on src/main/java/main/ProYapper.java:
```diff
@@ -0,0 +1,36 @@
+package main;
```
I appreciate the separation of different types of classes into different packages! It keeps the project well organized. Good job! :)

sannie-beep commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    @Override
+    public String toString() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MMM dd yyyy, h:mm a");
```
Perhaps you consider making this into a static final constant, e.g. PREFERRED_DATE_TIME_FORMAT in the superclass Task, such that both Event and Deadline task can use it. This will ensure that the format you want is standardized across your package.

sannie-beep commented on src/main/java/tasks/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+     */
+    @Override
+    public String toFileString() {
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
Same as the previous. Since this is a different format from the one used during runtime, it would help to create two separate constants for both saving format and runtime format, within the task class (maybe SAVING_DATE_TIME_FORMAT and RUNTIME_DATE_TIME_FORMAT). But nice job differentiating the two formats using javadocs to help me understand better what each separate format is used for :)!

sannie-beep reviewed (commented):
Overall well done, organised and well-formatted! Nice job with the organisation into different packages for classes of different functionality. Well-explained using javadocs too. Just a few minor suggestions to improve standardisation with the date-time formatting, as well as the Ui message displaying for added and removed tasks. Good job! :)

TechRaj committed:
Fix dialog box cutting off part of long text

Ensure that long text in the dialog box is fully displayed by allowing the VBox height to adjust dynamically. This resolves an issue where the text was truncated if it exceeded the height of the dialog box.

Tested with various input lengths to confirm the fix.

TechRaj committed:
Fix dialog box cutting off part of long text

Ensure that long text in the dialog box is fully displayed by allowing the VBox height to adjust dynamically. This resolves an issue where the text was truncated if it exceeded the height of the dialog box.

Tested with various input lengths to confirm the fix.

TechRaj committed:
Fix dialog box cutting off part of long text

Ensure that long text in the dialog box is fully displayed by allowing the VBox height to adjust dynamically. This resolves an issue where the text was truncated if it exceeded the height of the dialog box.

Tested with various input lengths to confirm the fix.

TechRaj committed:
Merge branch 'branch-test' into branch-A-Assertions

* branch-test:
  Fix dialog box cutting off part of long text

TechRaj committed:
Add assertions across multiple classes to ensure input validity

This commit introduces assertions in several classes to validate inputs and
prevent potential issues. The assertions ensure that invalid inputs do not
cause unexpected behavior during runtime.

Key changes include:
- TaskList: Added assertions to check for non-null tasks and valid indices
  in addTask, deleteTask, and get methods.
- Ui: Added assertions to validate non-null message inputs before displaying
  messages or errors.
- Parser: Added assertions to ensure valid command parsing.

These changes improve the robustness of the system by catching potential
errors early in the program execution.

TechRaj committed:
Merge pull request #1 from TechRaj/branch-A-Assertions

Assertions

TechRaj committed:
Improve code quality across the application

Refactor several classes to improve code readability, maintainability, and adherence to coding standards:

- Extracted complex logic into smaller methods in the `Storage` class to reduce arrowhead anti-pattern.
- Introduced method-specific validation checks to ensure better error handling.
- Improved consistency with naming conventions and encapsulation throughout.
- Added missing Javadoc comments to ensure comprehensive documentation.
- Ensured that wildcards in imports are removed and replaced with explicit class imports to improve clarity.

These changes improve the overall structure and readability of the code, making it easier to maintain and extend in the future.

TechRaj committed:
Merge pull request #2 from TechRaj/branch-A-CodeQuality

Improve code quality across the application

TechRaj committed:
Add priority filtering feature to the chatbot

- Implemented a new `PriorityCommand` that allows users to filter tasks by their priority level (high, medium, or low).
- Added `parsePriorityCommand` in the `Parser` class to correctly handle the `priority` command input.
- Enhanced validation in the `parsePriorityCommand` to ensure only valid priority levels (high, medium, or low) are accepted.
- Updated the `TaskList` to filter tasks based on the specified priority.
- Integrated priority filtering into the command execution flow and ensured that results are properly displayed via the UI.

This new feature allows users to easily view tasks based on their urgency or importance, improving task management and organisation within the chatbot.

TechRaj committed:
Merge pull request #3 from TechRaj/branch-C-Priority

Add priority filtering feature to the chatbot

TechRaj committed:
Improve code quality and enhance UI with emojis

- Refactored the MainWindow class to remove arrowhead anti-pattern, improving readability and maintainability.
- Applied more SLAP (Single Level of Abstraction Principle) across various classes, breaking down complex methods into smaller, more manageable ones.
- Added more emojis to the UI to enhance user-friendliness and make the task list and chatbot responses more visually engaging.
- Ensured that the refactoring didn't affect existing functionality and improved the overall structure for future scalability.

TechRaj committed:
Add detailed user guide to README

- Added sections on task management including deadlines, todos, and events.
- Included examples and expected outcomes for adding tasks, listing tasks, and filtering by priority.
- Added instructions for marking tasks as done/not done and finding specific tasks.
- Documented task deletion process with examples and expected output.
- Enhanced clarity with structured examples and descriptions of expected outcomes for each command.

TechRaj committed:
Merge branch 'master' of https://github.com/TechRaj/ip

* 'master' of https://github.com/TechRaj/ip:

TechRaj committed:
Update priority icon handling to adapt to different OS environments

- Modified `getPriorityIcon()` method to check the operating system
  and adjust the priority icons accordingly.
- On macOS, emoji icons (🔴, 🟡, 🟢) are used for high, medium, and
  low priorities respectively.
- On other OSes, text-based icons ([HIGH], [MED], [LOW]) are used to
  ensure compatibility and clarity.
- Updated `toString()` methods in `Task`, `Deadline`, `ToDo`, and
  `Event` classes to use the new priority icon handling.

This update improves cross-platform compatibility by ensuring that
priority icons are displayed consistently on both macOS and Windows.

TechRaj committed:
Update README with OS-specific task priority display examples

- Added a section in the README to show how task priorities will
  appear differently on macOS and Windows.
- Provided examples of output using emoji icons (macOS) and text
  icons ([HIGH], [MED], [LOW]) for Windows users.
- Clarified the cross-platform behavior of the priority feature to
  ensure users are aware of how tasks will be displayed on different
  operating systems.

This update helps users understand how the priority icons will be
rendered based on the operating system they are using.
